src/nifs/tests.rs

```diff
@@ -0,0 +1,683 @@
+    let n_f64 = n as f64;
+    let mul_res = n_f64 * (2f64.powi(K as i32));
+    let log_result = mul_res.log2().ceil();
+    log_result as usize
+}
+
+// test with single custom gate without lookup
+mod zero_round_test {
+    use super::*;
+    use crate::main_gate::{MainGate, MainGateConfig, RegionCtx};
+
+    const T: usize = 3;
+
+    #[derive(Clone, Debug)]
+    struct TestCircuitConfig {
+        pconfig: MainGateConfig<T>,
+        instance: Column<Instance>,
+    }
+
+    struct TestCircuit<F: PrimeField> {
+        inputs: Vec<F>,
+        r: F,
+    }
+
+    impl<F: PrimeField> TestCircuit<F> {
+        fn new(inputs: Vec<F>, r: F) -> Self {
+            Self { inputs, r }
+        }
+    }
+
+    impl<F: PrimeField + FromUniformBytes<64>> Circuit<F> for TestCircuit<F> {
+        type Config = TestCircuitConfig;
+        type FloorPlanner = SimpleFloorPlanner;
+
+        fn without_witnesses(&self) -> Self {
+            Self {
+                inputs: Vec::new(),
+                r: F::ZERO,
+            }
+        }
+
+        fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+            let instance = meta.instance_column();
+            meta.enable_equality(instance);
+            Self::Config {
+                pconfig: MainGate::configure(meta),
+                instance,
+            }
+        }
+
+        fn synthesize(
+            &self,
+            config: Self::Config,
+            mut layouter: impl Layouter<F>,
+        ) -> Result<(), Error> {
+            let pchip = MainGate::new(config.pconfig);
+            let output = layouter.assign_region(
+                || "test",
+                |region| {
+                    let ctx = &mut RegionCtx::new(region, 0);
+                    pchip.random_linear_combination(ctx, self.inputs.clone(), self.r)
+                },
+            )?;
+            layouter.constrain_instance(output.cell(), config.instance, 0)?;
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn test_nifs() -> Result<(), NIFSError> {
+        const K: u32 = 4;
+        let inputs1 = (1..10).map(Fr::from).collect();
+        let inputs2 = (2..11).map(Fr::from).collect();
+        let circuit1 = TestCircuit::new(inputs1, Fr::from_u128(2));
+        let output1 = Fr::from_u128(4097);
+        let public_inputs1 = vec![output1];
+        let mut td1 = TableData::<Fr>::new(K, public_inputs1);
+        let _ = td1.assembly(&circuit1);
+
+        let circuit2 = TestCircuit::new(inputs2, Fr::from_u128(3));
+        let output2 = Fr::from_u128(93494);
+        let public_inputs2 = vec![output2];
+        let mut td2 = TableData::<Fr>::new(K, public_inputs2);
+        let _ = td2.assembly(&circuit2);
+
+        let p1 = smallest_power(td1.cs.num_advice_columns(), K);
+        let p2 = smallest_power(td1.cs.num_selectors() + td1.cs.num_fixed_columns(), K);
+        let ck = CommitmentKey::<G1Affine>::setup(p1.max(p2), b"zero_round_test");
+
+        fold_instances(&ck, &td1, &td2)
+    }
```

Comment:
In a past comment, I provided a refactoring of this method that allowed you to see all the cases in one glance. There were no extra variables, those that were there were near the point of use, and assembly was checked for execution. I can do this refactoring myself for all `test_nifs`, but this version is very hard to read and debug.

Reply:
can you elaborate more on "There were no extra variables, those that were there were near the point of use, and assembly was checked for execution."

Actually, all the test and debug logic is in `fold_instances` and not in previous lines (which is to create two circuit instances to be folded)

Reply:
And if we decide to refactor `test_nifs`, we can do it in a separate PR.

Reply:
Made in separate PR #93